src/crypto/merkle.rs

```diff
@@ -93,36 +97,58 @@ const EMPTY_ROOTS: [Hash; MAX_TREE_HEIGHT] = [
     hex!("33ebfb34d3aa119cb665d564acdd77b318dc86aa6a97744edb3bc03e97d776ca"),
 ];
 
+/// Marker trait for the leaf nodes of a Merkle tree.
+pub trait MerkleLeaf: AsRef<[u8]> {}
+
+/// Marker trait for the root of a Merkle tree.
+pub trait MerkleRoot: From<Hash> + Into<Hash> + Clone {}
+
+/// Marker trait for the proof of a Merkle tree.
+pub trait MerkleProof: AsRef<[Hash]> + From<Vec<Hash>> {}
+
+impl MerkleLeaf for Vec<u8> {}
+impl MerkleRoot for Hash {}
+impl MerkleProof for Vec<Hash> {}
+
+pub type PlainMerkleTree = MerkleTree<Vec<u8>, Hash, Vec<Hash>>;
+pub type SliceMerkleTree = MerkleTree<Vec<u8>, Hash, Vec<Hash>>;
+pub type DoubleMerkleTree = MerkleTree<Hash, Hash, Vec<Hash>>;
+
 /// Implementation of a Merkle tree.
-pub struct MerkleTree {
+pub struct MerkleTree<L: AsRef<[u8]>, R: MerkleRoot, P: MerkleProof> {
```

Comment:
Shouldn't this be the following?

```suggestion
pub struct MerkleTree<L: MerkleLeaf, R: MerkleRoot, P: MerkleProof> {
```

This will require a couple of other minor changes to compile.